### PR TITLE
add injection support to junit4

### PR DESCRIPTION
**Pull request by rotty3000**
.. and a whole lot more. I basically refactored the entire BundleContext and Service support to use injection in both JUnit 4 and 5. There are README's for both and hopefully you'll note a significantly improved usage.

Signed-off-by: Raymond Augé <raymond.auge@liferay.com>

### CODE REVIEW

**Comment by rotty3000** [comment]
I did add many more tests including the missing negative tests for cardinality and timeout but also many more.

**Comment by rotty3000** [comment]
.. Hopefully you don't find the FieldInjector too _borrowy_ (from junit5)!

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	public ServiceRule			sur	= new ServiceRule();
+
+	@InjectService(filter = "(%s=%s)", filterArguments = {
+		"objectClass", "org.osgi.service.log.LoggerFactory"
+	})
+	LogService					logService;
+	@InjectService(filter = "(%s=%s)", filterArguments = {
+		"objectClass", "org.osgi.service.log.LoggerFactory"
+	})
+	BaseServiceUse<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java:39]: Shouldn't the field type be `ServiceAware<LogService>`?

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) OSGi Alliance (2019). All Rights Reserved.
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java:2]: All these new source files should just be `(2020)`.

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+public class FilterTest {
+
+	@Rule
+	public ServiceRule			sur	= new ServiceRule();
+
+	@InjectService(filter = "(objectClass=org.osgi.service.log.LoggerFactory)")
+	LogService					logService;
+	@InjectService(filter = "(objectClass=org.osgi.service.log.LoggerFactory)")
+	BaseServiceUse<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java:35]: Shouldn't the field type be `ServiceAware<LogService>`?

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/MultiCardinalityServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.common.dictionary.Dictionaries;
+import org.osgi.test.common.service.ServiceAware;
+import org.osgi.test.junit4.types.Foo;
+
+public class MultiCardinalityServiceTest {
+
+	private static List<ServiceRegistration<?>>	regisrations	= new CopyOnWriteArrayList<>();
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/MultiCardinalityServiceTest.java:38]: variable name spelling.

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/OwnServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+public class OwnServiceTest {
+
+	@Rule
+	public ServiceRule			sur	= new ServiceRule();
+
+	@InjectBundleContext
+	BundleContext				bundleContext;
+	@InjectService(cardinality = 0)
+	ServiceAware<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/OwnServiceTest.java:38]: I would assume this test should be looking for `ServiceAware<Foo>` since it is named `OwnServiceTest` and the test registers a `Foo` service.

File: org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceConfiguration.java [hunk 6]
```diff
@@ -54,7 +54,8 @@ public void close() throws Exception {
 
 	@Override
 	public String toString() {
-		return String.format("ServiceUseExtension [Class=\"%s\", filter=\"%s\", cardinality=%s, timeout=%s]",
+		return String.format(
+			"ServiceUseConfiguration [Class=\"%s\", filter=\"%s\", cardinality=%s, timeout=%s]",
```
Comment (bjhargrave) [inline org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceConfiguration.java:58]: Name should be `ServiceConfiguration`.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+public class FilterArgumentsTest {
+
+	@InjectService(filter = "(%s=%s)", filterArguments = {
+		"objectClass", "org.osgi.service.log.LoggerFactory"
+	})
+	LogService					logService;
+	@InjectService(filter = "(%s=%s)", filterArguments = {
+		"objectClass", "org.osgi.service.log.LoggerFactory"
+	})
+	BaseServiceUse<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java:37]: `ServiceAware<LogService>`

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.common.service.BaseServiceUse;
+
+@ExtendWith(ServiceExtension.class)
+public class FilterTest {
+
+	@InjectService(filter = "(objectClass=org.osgi.service.log.LoggerFactory)")
+	LogService					logService;
+	@InjectService(filter = "(objectClass=org.osgi.service.log.LoggerFactory)")
+	BaseServiceUse<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java:33]: `ServiceAware<LogService>`

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/OwnServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import org.osgi.test.junit5.types.Foo;
+
+@ExtendWith(BundleContextExtension.class)
+@ExtendWith(ServiceExtension.class)
+public class OwnServiceTest {
+
+	@InjectBundleContext
+	BundleContext				bundleContext;
+	@InjectService(cardinality = 0)
+	ServiceAware<LogService>	lsServiceAware;
```
Comment (bjhargrave) [inline org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/OwnServiceTest.java:38]: `ServiceAware<Foo>`

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java (hunk 1 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java:39]: So provided the type is at least ServiceAware you can use it. :) there are a couple of our higher order types that have some added methods. I'll document those in another pass. For instance, this type let's me introspect the `ServiceConfiguration` details which I'm using the assert function.

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java (hunk 2 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java:2]: right right.. fixing

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java (hunk 3 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java:35]: same as above.

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/OwnServiceTest.java (hunk 5 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/OwnServiceTest.java:38]: nice catch... :|

File: org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceConfiguration.java (hunk 6 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceConfiguration.java:58]: fixing

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java (hunk 7 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java:37]: see earlier comment.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java (hunk 8 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java:33]: see earlier comment.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/OwnServiceTest.java (hunk 9 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/OwnServiceTest.java:38]: fixing

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java (hunk 1 above)
Comment (bjhargrave) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java:39]: Using BaseServiceUse is implementation specific. You just happen to know it will work. But if I implemented ServiceAware or subclassed ServiceAware and tried to use that, the injection code would see type is assignable but fail to set the field. I am not sure using is assignable in the injection code is wise since we cannot support any subtype of ServiceAware (or List).

**Comment by rotty3000** [comment]
Updated!

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java (hunk 8 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterTest.java]: fixed

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java (hunk 7 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/FilterArgumentsTest.java]: fixed

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java (hunk 3 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterTest.java]: fixed

File: org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java (hunk 1 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/service/FilterArgumentsTest.java]: fixed, now ckecks and expects exactly the types supported.

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+ * &#64;InjectService
+ * Foo foo;
+ *
+ * &#64;Test
+ * public void aTest() {
+ * 	// use foo
+ * }
+ * </pre>
+ */
+public class ServiceRule extends BundleContextRule implements AutoCloseable, MethodRule {
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:55]: It seems odd to me that ServiceRule extends BundleContextRule. If that is necessary, then perhaps we only need one JUnit 4 rule?

File: org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java [hunk 11]
```diff
@@ -21,9 +21,10 @@
 
 import org.osgi.annotation.versioning.ProviderType;
 import org.osgi.framework.ServiceReference;
+import org.osgi.test.common.tracking.TrackingConfig;
 
 @ProviderType
-public interface ServiceAware<T> {
+public interface ServiceAware<T> extends TrackingConfig {
```
Comment (bjhargrave) [inline org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java:27]: I know this extends helps you testing but it seems the wrong abstraction. For the testing can't we just inject a ServiceAware object and cast to BaseServiceUse or TrackingConfig since we know about our impl?

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java (hunk 10 above)
Comment (kriegfrj) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:55]: Unlike JUnit 5 Extensions which have the store,  there's no built-in mechanism in JUnit 4 rules to synchronize between rules. I'm guessing he's fine this to make the synchronization easier. But maybe composition would be better than inheritance?

File: org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java (hunk 11 above)
Comment (rotty3000) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java:27]: There was no reason to hide this interface from the end user. In fact it contains details that are quite useful for things like the assertj's diagnostic `.as(...)' function.

Comment (bjhargrave) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java:27]: > There was no reason to hide this interface from the end user. In fact it contains details that are quite useful for things like the assertj's diagnostic `.as(...)' function.

OK, Then why not just put those methods in ServiceAware and get rid of TrackingConfig? I don't see any separate need for TrackingConfig then.

Comment (rotty3000) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java:27]: because they are implemented at different layers.

Comment (bjhargrave) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java:27]: I don't see that. TrackServices implements it for no real reason. BaseServiceUse implements it (as well as ServiceAware) and delegates the methods to a TrackServices instance. So if you just move the methods from TrackingConfig to ServiceAware, you do not need TrackingConfig.

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java (hunk 10 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java]: I moved it back to composition but you'll note that 99% of the time the default constructor will be used, anyway, it's not a big issue.

File: org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java (hunk 11 above)
Comment (rotty3000) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java]: I pushed a change with cast instead.

**Comment by rotty3000** [comment]
@bjhargrave I eliminated the TrackerConfig interface as suggested.

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	@Override
+	public void close() throws Exception {
+		for (Iterator<Entry<InjectService, ServiceConfiguration<?>>> itr = configurations.entrySet()
+			.iterator(); itr.hasNext();) {
+			Entry<InjectService, ServiceConfiguration<?>> entry = itr.next();
+			entry.getValue()
+				.close();
+			itr.remove();
+		}
+		bundleContextRule.close();
```
Comment (bjhargrave) [inline org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: This may be a problem if the `bundleContextRule` object was supplied to the `ServiceRule(BundleContextRule)` constructor. That is, should our close result in closing a supplied BundleContextRule? If we created the BundleContextRule, as is done by the `ServiceRule()` constructor, then we obviously want to close it. 

Perhaps the simple answer is to get rid of the `ServiceRule(BundleContextRule)` constructor. We just always create a new BundleContextRule. I don't see any use of the `ServiceRule(BundleContextRule)` constructor in this PR.

Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: that depends on if we consider the consequences of having:
```
@Rule BundleContextRule bcr = new BundleContextRule();
@Rule ServiceRule sr = new ServiceRule();
```
where now you have double injection happening.
as opposed to having:
```
@Rule BundleContextRule bcr = new BundleContextRule();
@Rule ServiceRule sr = new ServiceRule(bcr);
```
where the internal state of BundleContextRule is managed and avoids the double injection

Comment (bjhargrave) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: Fine. But then `sr.close()` must not close `bcr`.

Comment (bjhargrave) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: And what if JUnit calls `bcr.close()` before `sr.close()`? Will 'sr.close()` fail because of its dependence on a closed `bcr`?

Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: did you look at the code of BCR close? it should be resonably implemented so it doesn't matter who calls it first?

Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: it shouldn't matter if bcr is closed I don't think

Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: ... this is why I had used inheritance...

Comment (bjhargrave) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: It looks like this will call ServiceTracker.close which will call BundleContext.ungetService for each tracked service. Assuming the real BundleContext is still valid, then no issue.

Comment (bjhargrave) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: >  this is why I had used inheritance...

That just means you always had your own copy of a BundleContextRule which is what I was saying by removing the ServiceRule(BundleContextRule) constructor. Same result with aggregation instead of inheritance. Inheritance is a problem since it limits future implementation options.

Comment (bjhargrave) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java:98]: A ServiceRule _is not_ a BundleContextRule. A ServiceRule just needs a BundleContext.

**Comment by rotty3000** [comment]
more refactor

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java (hunk 12 above)
Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java]: fixed all this
